cmd/pdc/main.go

```diff
@@ -2,21 +2,165 @@ package main
 
 import (
 	"context"
+	"flag"
+	"fmt"
 	"os"
 	"os/signal"
 	"syscall"
 
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+
 	"github.com/grafana/dskit/services"
+	"github.com/grafana/pdc-agent/pkg/pdc"
 	"github.com/grafana/pdc-agent/pkg/ssh"
 )
 
+type mainFlags struct {
+	PrintHelp bool
+	LogLevel  string
+}
+
+func (mf *mainFlags) RegisterFlags(fs *flag.FlagSet) {
+	fs.BoolVar(&mf.PrintHelp, "h", false, "Print help")
+	fs.StringVar(&mf.LogLevel, "log.level", "info", `"debug", "info", "warn" or "error"`)
+}
+
 func main() {
+	sshConfig := ssh.DefaultConfig()
+	mf := &mainFlags{}
+	pdcClientCfg := &pdc.Config{}
+
+	sshConfig.Args = os.Args[1:]
+
+	legacyMode := inLegacyMode()
+
+	if legacyMode {
+		sshConfig.LegacyMode = true
+		err := runLegacyMode(sshConfig)
+		if err != nil {
+			fmt.Printf("error: %s", err)
+			os.Exit(1)
+		}
+		return
+	}
+
+	usageFn, err := parseFlags(mf.RegisterFlags, sshConfig.RegisterFlags, pdcClientCfg.RegisterFlags)
+	if err != nil {
+		fmt.Println("cannot parse flags")
+		os.Exit(1)
+	}
+
+	logger := setupLogger(mf.LogLevel)
+
+	if mf.PrintHelp {
+		usageFn()
+		return
+	}
+
+	err = run(logger, sshConfig, pdcClientCfg)
+	if err != nil {
+		level.Error(logger).Log("err", err)
+		os.Exit(1)
+	}
+
+	sshConfig.PDC = *pdcClientCfg
+
+}
+
+func run(logger log.Logger, sshConfig *ssh.Config, pdcConfig *pdc.Config) error {
 	ctx, stop := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
 	defer stop()
+
+	pdcClient, err := pdc.NewClient(pdcConfig, logger)
+	if err != nil {
+		level.Error(logger).Log("msg", fmt.Sprintf("cannot initialise PDC client: %s", err))
+		return err
+	}
+
+	// Whilst KeyManager is not passed to ssh.Client, we need KM to have run before ssh.Client is running.
```

Comment:
I think we should pass the km to the ssh client 🤔 . The SSH certificate is only valid for ~1 hour, so we should ensure a new certificate is issued each time the SSH client gets disconnected ([here](https://github.com/grafana/pdc-agent/blob/2b369bfb3d3df55f15247cfc67e769fc673a786d/pkg/ssh/ssh.go#L122-L126))

Reply:
I quickly tried locally. Indeed the current code breaks after an hour (`killall ssh` to test). This can be fixed by calling km.EnsureCertExists in the [loop](https://github.com/grafana/pdc-agent/blob/2b369bfb3d3df55f15247cfc67e769fc673a786d/pkg/ssh/ssh.go#L122-L126))

Reply:
Great, this came to me over the weekend as well. I'll address this!